File: src/main/java/graphql/relay/Relay.java
package graphql.relay;


import graphql.schema.*;

import java.util.ArrayList;
import java.util.List;

import static graphql.Scalars.*;
import static graphql.schema.GraphQLArgument.newArgument;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;
import static graphql.schema.GraphQLInputObjectField.newInputObjectField;
import static graphql.schema.GraphQLInputObjectType.newInputObject;
import static graphql.schema.GraphQLInterfaceType.newInterface;
import static graphql.schema.GraphQLObjectType.newObject;

public class Relay
{

    public static final String NODE = "Node";
    private GraphQLObjectType pageInfoType = Companion.newObject()
                                                      .name("PageInfo")
                                                      .description("Information about pagination in a connection.")
                                                      .field(Companion.newFieldDefinition()
                                                                      .name("hasNextPage")
                                                                      .type(new GraphQLNonNull(INSTANCE.getGraphQLBoolean()))
                                                                      .description("When paginating forwards, are there more items?"))
                                                      .field(Companion.newFieldDefinition()
                                                                      .name("hasPreviousPage")
                                                                      .type(new GraphQLNonNull(INSTANCE.getGraphQLBoolean()))
                                                                      .description("When paginating backwards, are there more items?"))
                                                      .field(Companion.newFieldDefinition()
                                                                      .name("startCursor")
                                                                      .type(INSTANCE.getGraphQLString())
                                                                      .description("When paginating backwards, the cursor to continue."))
                                                      .field(Companion.newFieldDefinition()
                                                                      .name("endCursor")
                                                                      .type(INSTANCE.getGraphQLString())
                                                                      .description("When paginating forwards, the cursor to continue."))
                                                      .build();

    public GraphQLInterfaceType nodeInterface(TypeResolver typeResolver)
    {
        GraphQLInterfaceType node = Companion.newInterface()
                                             .name(NODE)
                                             .description("An object with an ID")
                                             .typeResolver(typeResolver)
                                             .field(Companion.newFieldDefinition()
                                                             .name("id")
                                                             .description("The ID of an object")
                                                             .type(new GraphQLNonNull(INSTANCE.getGraphQLID())))
                                             .build();
        return node;
    }

    public GraphQLFieldDefinition nodeField(GraphQLInterfaceType nodeInterface, DataFetcher nodeDataFetcher)
    {
        GraphQLFieldDefinition fieldDefinition = Companion.newFieldDefinition()
                                                          .name("node")
                                                          .description("Fetches an object given its ID")
                                                          .type(nodeInterface)
                                                          .dataFetcher(nodeDataFetcher)
                                                          .argument(Companion.newArgument()
                               .name("id")
                               .description("The ID of an object")
                               .type(new GraphQLNonNull(INSTANCE.getGraphQLID())))
                                                          .build();
        return fieldDefinition;
    }

    public List<GraphQLArgument> getConnectionFieldArguments()
    {
        List<GraphQLArgument> args = new ArrayList<GraphQLArgument>();

        args.add(Companion.newArgument()
                          .name("before")
                          .type(INSTANCE.getGraphQLString())
                          .build());
        args.add(Companion.newArgument()
                          .name("after")
                          .type(INSTANCE.getGraphQLString())
                          .build());
        args.add(Companion.newArgument()
                          .name("first")
                          .type(INSTANCE.getGraphQLInt())
                          .build());
        args.add(Companion.newArgument()
                          .name("last")
                          .type(INSTANCE.getGraphQLInt())
                          .build());
        return args;
    }

    public List<GraphQLArgument> getBackwardPaginationConnectionFieldArguments()
    {
        List<GraphQLArgument> args = new ArrayList<GraphQLArgument>();

        args.add(Companion.newArgument()
                          .name("before")
                          .type(INSTANCE.getGraphQLString())
                          .build());
        args.add(Companion.newArgument()
                          .name("last")
                          .type(INSTANCE.getGraphQLInt())
                          .build());
        return args;
    }

    public List<GraphQLArgument> getForwardPaginationConnectionFieldArguments()
    {
        List<GraphQLArgument> args = new ArrayList<GraphQLArgument>();

        args.add(Companion.newArgument()
                          .name("after")
                          .type(INSTANCE.getGraphQLString())
                          .build());
        args.add(Companion.newArgument()
                          .name("first")
                          .type(INSTANCE.getGraphQLInt())
                          .build());
        return args;
    }

    public GraphQLObjectType edgeType(String name,
                                      GraphQLOutputType nodeType,
                                      GraphQLInterfaceType nodeInterface,
                                      List<GraphQLFieldDefinition<?>> edgeFields)
    {

        GraphQLObjectType edgeType = Companion.newObject()
                                              .name(name + "Edge")
                                              .description("An edge in a connection.")
                                              .field(Companion.newFieldDefinition()
                                                              .name("node")
                                                              .type(nodeType)
                                                              .description("The item at the end of the edge"))
                                              .field(Companion.newFieldDefinition()
                                                              .name("cursor")
                                                              .type(new GraphQLNonNull(INSTANCE.getGraphQLString()))
                                                              .description(""))
                                              .fields(edgeFields)
                                              .build();
        return edgeType;
    }

    public GraphQLObjectType connectionType(String name,
                                            GraphQLObjectType edgeType,
                                            List<GraphQLFieldDefinition<?>> connectionFields)
    {

        GraphQLObjectType connectionType = Companion.newObject()
                                                    .name(name + "Connection")
                                                    .description("A connection to a list of items.")
                                                    .field(Companion.newFieldDefinition()
                                                                    .name("edges")
                                                                    .type(new GraphQLList(edgeType)))
                                                    .field(Companion.newFieldDefinition()
                                                                    .name("pageInfo")
                                                                    .type(new GraphQLNonNull(pageInfoType)))
                                                    .fields(connectionFields)
                                                    .build();
        return connectionType;
    }


    public GraphQLFieldDefinition mutationWithClientMutationId(String name, String fieldName,
                                                               List<GraphQLInputObjectField> inputFields,
                                                               List<GraphQLFieldDefinition<?>> outputFields,
                                                               DataFetcher dataFetcher)
    {
        GraphQLInputObjectType inputObjectType = Companion.newInputObject()
                                                          .name(name + "Input")
                                                          .field(Companion.newInputObjectField()
                            .name("clientMutationId")
                            .type(new GraphQLNonNull(INSTANCE.getGraphQLString())))
                                                          .fields(inputFields)
                                                          .build();
        GraphQLObjectType outputType = Companion.newObject()
                                                .name(name + "Payload")
                                                .field(Companion.newFieldDefinition()
                                                                .name("clientMutationId")
                                                                .type(new GraphQLNonNull(INSTANCE.getGraphQLString())))
                                                .fields(outputFields)
                                                .build();

        return Companion.newFieldDefinition()
                        .name(fieldName)
                        .type(outputType)
                        .argument(Companion.newArgument()
                               .name("input")
                               .type(new GraphQLNonNull(inputObjectType)))
                        .dataFetcher(dataFetcher)
                        .build();
    }

    public static class ResolvedGlobalId
    {

        public ResolvedGlobalId(String type, String id)
        {
            this.type = type;
            this.id = id;
        }

        /**
         * @deprecated use {@link #getType()}
         */
        @Deprecated
        public String type;
        /**
         * @deprecated use {@link #getId()}
         */
        @Deprecated
        public String id;

        public String getType()
        {
            return type;
        }

        public String getId()
        {
            return id;
        }
    }

    public String toGlobalId(String type, String id)
    {
        return Base64.toBase64(type + ":" + id);
    }

    public ResolvedGlobalId fromGlobalId(String globalId)
    {
        String[] split = Base64.fromBase64(globalId).split(":", 2);
        return new ResolvedGlobalId(split[0], split[1]);
    }
}

File: src/main/java/graphql/relay/Relay.kt
package graphql.relay


import graphql.GraphQLBoolean
import graphql.GraphQLID
import graphql.GraphQLInt
import graphql.GraphQLString
import graphql.schema.*
import kotlin.properties.Delegates


private val pageInfoType = newObject {
    name = "PageInfo"
    description = "Information about pagination in a connection."
    field<Boolean> {
        name = "hasNextPage"
        description = "When paginating forwards, are there more items?"
        type = GraphQLNonNull(GraphQLBoolean)
    }
    field<Boolean> {
        name = "hasPreviousPage"
        description = "When paginating backwards, are there more items?"
        type = GraphQLNonNull(GraphQLBoolean)
    }
    field<String> {
        name = "startCursor"
        description = "When paginating backwards, the cursor to continue."
    }
    field<String> {
        name = "endCursor"
        description = "When paginating forwards, the cursor to continue."
    }
}

internal fun nodeInterface(resolver: TypeResolver) = newInterface {
    name = "Node"
    description = "An object with an ID"
    typeResolver = resolver
    field<String> {
        name = "id"
        description = "The ID of an object"
        type = GraphQLNonNull(GraphQLID)
    }
}

internal inline fun <reified T : Any> nodeField(nodeInterface: GraphQLInterfaceType, noinline nodeDataFetcher: DataFetcher<T>): GraphQLFieldDefinition<*> {
    return newField<T> {
        name = "node"
        description = "Fetches an object given its ID"
        type = nodeInterface
        fetcher = nodeDataFetcher
        argument {
            name = "id"
            description = "The ID of an object"
            type = GraphQLNonNull(GraphQLID)
        }
    }
}

val connectionFieldArguments: List<GraphQLArgument>
    get() {
        return listOf(newArgument().name("before").type(GraphQLString).build(),
                      newArgument().name("after").type(GraphQLString).build(),
                      newArgument().name("first").type(GraphQLInt).build(),
                      newArgument().name("last").type(GraphQLInt).build())
    }

val backwardPaginationConnectionFieldArguments: List<GraphQLArgument>
    get() {
        return listOf(newArgument().name("before").type(GraphQLString).build(),
                      newArgument().name("last").type(GraphQLInt).build())
    }

val forwardPaginationConnectionFieldArguments: List<GraphQLArgument>
    get() {
        return listOf(newArgument().name("after").type(GraphQLString).build(),
                      newArgument().name("first").type(GraphQLInt).build())
    }

@GraphQLDslMarker
class EdgeTypeBuilder {
    var baseName: String by Delegates.notNull<String>()
    val fields: MutableList<GraphQLFieldDefinition<*>> = mutableListOf()
    var nodeType: GraphQLOutputType by Delegates.notNull<GraphQLOutputType>()

    inline fun <reified T : Any> field(block: GraphQLFieldDefinition.Builder<T>.() -> Unit) {
        this.fields += newField(block)
    }
}

inline fun <reified T : Any> edgeType(block: EdgeTypeBuilder.() -> Unit) : GraphQLObjectType {
    val builder = EdgeTypeBuilder()
    builder.block()

    return newObject {
            name = builder.baseName + "Edge"
            description = "An edge in a connection."
            field<T> {
                name = "node"
                description = "The item at the end of the edge"
                type = builder.nodeType
            }
            field<String> {
                name = "cursor"
                type = GraphQLNonNull(GraphQLString)
            }
            fields += builder.fields
        }
}


@GraphQLDslMarker
class ConnectionTypeBuilder {
    var baseName: String by Delegates.notNull<String>()
    var edgeType: GraphQLOutputType by Delegates.notNull<GraphQLOutputType>()
    val fields: MutableList<GraphQLFieldDefinition<*>> = mutableListOf()

    inline fun <reified T : Any> field(block: GraphQLFieldDefinition.Builder<T>.() -> Unit) {
        this.fields += newField(block)
    }

}

fun <T> connectionType(block: ConnectionTypeBuilder.() -> Unit): GraphQLObjectType {
    val builder = ConnectionTypeBuilder()
    builder.block()

    return newObject {
        name = builder.baseName + "Connection"
        description = "A connection to a list of items."
        field<List<T>> {
            name = "edges"
            type = GraphQLList(builder.edgeType)
        }
        field<PageInfo> {
            name = "pageInfo"
            type = GraphQLNonNull(pageInfoType)
        }
        fields += builder.fields
    }
}


//inline fun <reified T : Any> mutationWithClientMutationId(baseName: String,
//                                                          fieldName: String,
//                                                          inputFields: List<GraphQLInputObjectField>,
//                                                          outputFields: List<GraphQLFieldDefinition<*>>,
//                                                          noinline dataFetcher: DataFetcher<T>): GraphQLFieldDefinition<*> {
//    val inputObjectType = newInputObject {
//        name = baseName + "Input"
//        field {
//            name = "clientMutationId"
//            type = GraphQLNonNull(GraphQLString)
//        }
//        fields += inputFields
//
//    }
//    val outputType = newObject {
//        name = baseName + "Payload"
//        field<String> {
//            name = "clientMutationId"
//            type = GraphQLNonNull(GraphQLString)
//        }
//        fields += outputFields
//    }
//
//    return newField <T> {
//        name = fieldName
//        type = outputType
//        argument {
//            name = "input"
//            type = GraphQLNonNull(inputObjectType)
//        }
//        fetcher = dataFetcher
//    }
//}